contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchingStrategy.java

```diff
@@ -0,0 +1,158 @@
+package com.linkedin.parseq.batching;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+import com.linkedin.parseq.Task;
+import com.linkedin.parseq.batching.BatchImpl.BatchBuilder;
+import com.linkedin.parseq.batching.BatchImpl.BatchEntry;
+import com.linkedin.parseq.internal.ContextImpl;
+import com.linkedin.parseq.internal.PlanContext;
+import com.linkedin.parseq.promise.CountDownPromiseListener;
+import com.linkedin.parseq.promise.PromiseListener;
+import com.linkedin.parseq.promise.Promises;
+import com.linkedin.parseq.promise.SettablePromise;
+import com.linkedin.parseq.trace.Relationship;
+import com.linkedin.parseq.trace.ShallowTraceBuilder;
+import com.linkedin.parseq.trace.TraceBuilder;
+
+public abstract class BatchingStrategy<G extends Group, K, T> {
+
+  private final ConcurrentHashMap<Long, BatchBuilder<K, T>> _batches =
+      new ConcurrentHashMap<>();
+
+  public Task<T> batchable(final String desc, final K key) {
```

Comment:
batchable() returns new task that may be batched by the strategy but this is what consumer gets.